usr.sbin/pppd/pppstats/pppstats.8
.\"	@(#) $Header: /usr/tmp/cvs2git/cvsroot-netbsd/src/usr.sbin/pppd/pppstats/Attic/pppstats.8,v 1.6 1996/08/10 21:25:45 explorer Exp $ (LBL)
.Dd November 15, 1994
.Dt PPPSTATS 8
.Sh NAME
.Nm pppstats
.Nd report statistics for the specified PPP interface
.Sh SYNOPSIS
.Nm pppstats
.Op Fl v
.Op Fl r
.Op Fl z
.Op Fl a
.Op Fl c Ar count
.Op Fl w Ar wait
.Op Ar interface
.Sh DESCRIPTION
.Pp
The 
.Nm pppstats
utility reports PPP-related statistics at regular intervals for the
specified PPP
.Ar interface .
If the
.Ar interface
is unspecified, it will default to ppp0.
The display is split horizontally
into input and output sections containing columns of statistics
describing the properties and volume of packets received and
transmitted by the interface.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl a
Display absolute values rather than deltas.  If this option is
specified with the
.Fl c
and/or
.Fl w
options, the second and subsequent reports (as well as the first) will
show statistics for the time since the link was initiated.
.It Fl c Ar count
Repeat the display
.Ar count
times. The first display is for the time since the link was initiated
(that is, since the corresponding
.Xr pppd 8
was started), and each
subsequent report is for the time period since the last display
(unless the
.Fl a
option is specified).
If this option is not specified, the default repeat
.Ar count
is 1 if the
.Fl w
option is not specified, otherwise infinity.
.It Fl r
Display additional statistics summarizing the compression ratio
achieved by the packet compression algorithm in use.
.It Fl v
Display additional statistics demonstrating the efficacy of VJ header
compression and providing more explicit information on the operation
of the algorithm.
.It Fl w Ar wait
Pause
.Ar wait
seconds between each display. If no
.Ar wait
interval is specified, the default is 5 seconds.
.It Fl z
Instead of the standard display, show statistics indicating the
performance of the packet compression algorithm in use.
.El
.Pp
The following fields are printed on the input side when the
.Fl z
option is not used:
.Bl -tag -width search
.It Li IN
The total number of bytes received by this interface.
.It Li PACK
The total number of packets received by this interface.
.It Li VJCOMP
The number of compressed TCP packets received by this interface.
.It Li VJUNC
The number of uncompressed TCP packets received by this interface.
Not reported when the
.Fl r
option is specified.
.It Li VJERR
The number of corrupted or bogus TCP packets received by this interface.
Not reported when the
.Fl r
option is specified.
.It Li VJTOSS
The number of VJ header-compressed TCP packets dropped on reception by
this interface.  Only reported when the
.Fl v
option is specified.
.It Li NON-VJ
The total number of non-TCP packets received by this interface. Only
reported when the
.Fl v
option is specified.
.It Li RATIO
The compression ratio achieved for received packets by the
packet compression scheme in use, defined as the uncompressed size
divided by the compressed size.
Only reported when the
.Fl r
option is specified.
.It Li UBYTE
The total number of bytes received, after decompression of compressed
packets.  Only reported when the
.Fl r
option is specified.
.El
.Pp
The following fields are printed on the output side:
.Bl -tag -width search
.It Li OUT
The total number of bytes transmitted from this interface.
.It Li PACK
The total number of packets transmitted from this interface.
.It Li VJCOMP
The number of TCP packets transmitted from this interface with
VJ-compressed TCP headers.
.It Li VJUNC
The number of TCP packets transmitted from this interface with
VJ-uncompressed TCP headers.
Not reported when the
.Fl r
option is specified.
.It Li NON-VJ
The total number of non-TCP packets transmitted from this interface.
Not reported when the
.Fl r
option is specified.
.It Li VJSRCH
The number of searches for the cached header entry for a VJ header
compressed TCP packet.  Only reported when the
.Fl v
option is specified.
.It Li VJMISS
The number of failed searches for the cached header entry for a
VJ header compressed TCP packet.  Only reported when the
.Fl v
option is specified.
.It Li RATIO
The compression ratio achieved for transmitted packets by the
packet compression scheme in use, defined as the size
before compression divided by the compressed size.
Only reported when the
.Fl r
option is specified.
.It Li UBYTE
The total number of bytes to be transmitted, before packet compression
is applied.  Only reported when the
.Fl r
option is specified.
.El
.Pp
When the
.Fl z
option is specified,
.Nm pppstats
instead displays the following fields, relating to the packet
compression algorithm currently in use.  If packet compression is not
in use, these fields will all display zeroes.  The fields displayed on
the input side are:
.Bl -tag -width search
.It Li COMPRESSED BYTE
The number of bytes of compressed packets received.
.It Li COMPRESSED PACK
The number of compressed packets received.
.It Li INCOMPRESSIBLE BYTE
The number of bytes of incompressible packets (that is, those which
were transmitted in uncompressed form) received.
.It Li INCOMPRESSIBLE PACK
The number of incompressible packets received.
.It Li COMP RATIO
The recent compression ratio for incoming packets, defined as the
uncompressed size divided by the compressed size (including both
compressible and incompressible packets).
.El
.Pp
The fields displayed on the output side are:
.Bl -tag -width search
.It Li COMPRESSED BYTE
The number of bytes of compressed packets transmitted.
.It Li COMPRESSED PACK
The number of compressed packets transmitted.
.It Li INCOMPRESSIBLE BYTE
The number of bytes of incompressible packets transmitted (that is,
those which were transmitted in uncompressed form).
.It Li INCOMPRESSIBLE PACK
The number of incompressible packets transmitted.
.It Li COMP RATIO
The recent compression ratio for outgoing packets.
.El
.Sh SEE ALSO
.Xr ppp 4 ,
.Xr pppd 8


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.3291 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 11.0 to NetBSD 12.0:

	macppc: Add drivebay(4), a driver to control the Xserve G4's drive bays
		via sysctl. [macallan 20250812]
	macppc: Added support for fan control on Xserve G4. [macallan 20250820]
	acpi(4): Updated ACPICA to 20250807. [christos 20250824]
	binutils: Import 2.45 and switch amd64 to it  [christos 20250826]
	gffb(4): Add backlight control to the GeForce framebuffer driver.
		[macallan 20250830]
	gdb: Import 16.3 and switch everyone to it [christos 20250902]
	openpam(3): update to 20250531 (zingiber) [christos 20250903]
	openldap: Import 2.6.10. [christos 20250904]
	unbound(8): Import 1.23.1. [christos 20250906]
	kernel: Add new memory barrier for paravirtualized drivers
		requiring store-before-load ordering. [riastradh 20250906]
	macppc: Add driver for the LED bar found on Xserve G4,
		controls lights depending on CPU usage. [macallan 20250908]
	m68k: Change softfloat build to libc/softfloat. [nat 20250910]
	fsck_ffs(8): Expand flags with -a and -i. [perseant 20250914]
	dumplfs(8): Add -n flag to allow dumping of a given inode from an
		inode block, or every inode in a block. [perseant 20250914]
	lfs: Add working in-kernel roll forward. [perseant 20250917]
	kernel: Add a shared function to query the common properties used
		for configuring an Ethernet address. [thorpej 20251004]
	gffb(4): Add backlight control for NVIDIA GeForce FX 5200 Go.
		[macallan 20251006]
	hypotl(3): Accommodate infinity for softfloat. [nat 20251007]
	arm, aarch64: Switch to GCC 14. [mrg 20251007]
	hppa: Switch to GCC 14. [mrg 20251007]
	riscv: Switch to GCC 14. [mrg 20251007]
	sparc, sparc64: Switch to GCC 14. [mrg 20251007]
	vax: Switch to GCC 14. [mrg 20251007]
	x86: Switch to GCC 14. [mrg 20251007]
	ucom(4): Add support for console over USB-to-serial adapter
		[manu 20251009]
	libc: Merge extended worker threads for POSIX asynchronous I/O
		(GSoC project by Ethan Miller). [christos 20251010]
	ssh(1), sshd(8): Import OpenSSH 10.2. [christos 20251011]
	gpt(8): Add various riscv-specific partition types. [thorpej 20251011]
	fsck_lfs(8): Enable checking LFS64 volumes. [perseant 20251012]
	m68k: New port: mac68ksf - Macintosh 68k softfloat. [nat 20251020]
	sc16is7xx(4): NXP SC16IX7xx family of UART chips that are behind
		a I2C or SPI bus.  [brad 20251024]
	compat_linux(8): Add support for semtimedop(2), from Ricardo Branco.
		[christos 20251026]
	ftp(1): Add "connect" as a synonym for "open". [lukem 20251028]
	ia64: Switch to GCC 14. [mrg 20251029]
	lfs: Add in-kernel cleaner. [perseant 20251106]
	m68k: Experimental new pmap module for systems with 68k MMUs
		(plus the HP MMU). [thorpej 20251108]
	compat_linux(8): Add prctl(2) lwp_{g,s}etname, from Kevin Bloom.
		[christos 20251110]
	sparc64: Add implementation of Sun's VIS instruction API.
		[nia 20251113]
	iconv(1): Support -o outfile like gnu does and check writes for
		failure. [christos 20251121]
	cgd(4): Improve throughput of AES disk encryption using SSE2.
		[riastradh 20251123]
	powerpc: Switch to GCC 14. [martin 20251128]
	dreamcast, atu(4), icp(4), gpsa(8): Relicense various NetBSD
		components from 4-clause to 3-clause BSD. [nia 20251128]
	umcpmio(4): Reorg and support for the MCP2210, USB to SPI
		bridge. [brad 20251129]
	ucom(4): Allow using as boot console. [manu 20251129]
	sparc64: Use 64-bit optimized AES code for disk encryption.
		[nia 20251201]
	iwm(4): Update driver code, improving performance and stability.
		[mlelstv 20251201]
	lfs: Various stability improvements. [perseant 20251202]
	viaide(4): Identify and attach VT6420 IDE (PATA) controller.
		[andvar 20251204]
	diff(1), diff3(1): Import BSD-licensed diff to replace GNU
		diffutils, build with MKBSDDIFF=yes. [nia 20251204]
	evbppc: Enable PaX security features in WII kernel configuration.
		[nia 20251204]
	amd64: Support ramdisks when booting in GENPVH mode.
		[khorben 20251205]
	ftp(1): Fix ASCII transfers with progress bar. [lukem 20251206]
	evbppc: wii: Add support for GameCube controller sockets
		as uhid(4) (now "userspace hid") devices.
		[jmcneill 20251208]
	mips: Switch to GCC 14. [mrg 20251209]
	openresolv(8): import 3.17.1 (patched to 3.17.3) [roy 20251213]
	editline(3): Add a method to set custom word characters,
		from bsiegert. [christos 20251214]
	evbarm: Add support for TI OMAP4 PandaBoard-ES, from Rui-Xiang Guo.
		[skrll 20251216]
	tzcode: Updated to 2025c. [christos 20251218]
	tzdata: Updated to 2025c (using 2025cgtz) [kre 20251219]
	re(4): Made the code MPSAFE with a private mutex.
		[mlelstv 20251220]
	lm(4): Support five aux fan sensors for NCT6794D. [isaki 20251224]
	inetd(8): Add an optional accept limit to stream/nowait services.
		[mlelstv 20251227]
	sun2: Add various MK build knobs and disable them on sun2 to trim
		down the installation. [thorpej 20251228]
	diff(1): Switch the default to MKBSDDIFF=yes. [nia 20260102]
	i386, amd64, sparc64: Exclude various base features from the
		CD-ROM installer and add checks that cause ISO image

usr.sbin/pppd/pppstats/pppstats.c
/*
 * print PPP interface statistics:
 *      pppstats [-v] [-c count] [-w wait] [-M core] [-N system ] [interface]
 * print SLIP interface statistics:
 *      slstats [-v] [-c count] [-w wait] [-M core] [-N system ] [interface]
 *
 *
 *	Brad Parker (brad@cayman.com) 6/92
 *
 * from the original "slstats" by Van Jacobson
 *
 * Copyright (c) 1989, 1990, 1991, 1992 Regents of the University of
 * California. All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted
 * provided that the above copyright notice and this paragraph are
 * duplicated in all such forms and that any documentation,
 * advertising materials, and other materials related to such
 * distribution and use acknowledge that the software was developed
 * by the University of California, Berkeley.  The name of the
 * University may not be used to endorse or promote products derived
 * from this software without specific prior written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 *
 *	Van Jacobson (van@ee.lbl.gov), Dec 31, 1989:
 *	- Initial distribution.
 */

#ifndef lint
/*static char rcsid[] =
    "@(#) $Header: /usr/tmp/cvs2git/cvsroot-netbsd/src/usr.sbin/pppd/pppstats/Attic/pppstats.c,v 1.8 1994/11/15 07:20:54 glass Exp $ (LBL)";*/
static char rcsid[] = "$Id: pppstats.c,v 1.8 1994/11/15 07:20:54 glass Exp $";
#endif

#include <sys/param.h>
#include <sys/mbuf.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <sys/file.h>

#include <net/if.h>
#include <netinet/in.h>
#include <netinet/in_systm.h>
#include <netinet/ip.h>
#include <netinet/ip_var.h>

#define	VJC	1		/* XXX */
#define INET			/* XXX */

#include <net/slcompress.h>
#include <net/if_slvar.h>
#include <net/if_ppp.h>

#include <ctype.h>
#include <errno.h>
#include <err.h>
#include <kvm.h>
#include <limits.h>
#include <nlist.h>
#include <paths.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define V(offset) ((line != 0) ? sc->offset - osc->offset : sc->offset)

#ifdef PPP
#define STRUCT			struct ppp_softc
#define NLIST_ENTRY		"_ppp_softc"
#define INTERFACE_PREFIX 	"ppp"
#else
#define STRUCT			struct sl_softc
#define NLIST_ENTRY 		"_sl_softc"
#define INTERFACE_PREFIX 	"sl"
#endif

struct nlist nl[] = {
#define N_SOFTC 0
        { NLIST_ENTRY },
	"",
};

kvm_t	*kd;
char	*progname, interface[IFNAMSIZ];
int	count, infinite, interval, signalled, unit, vflag;

void
usage()
{
	fprintf(stderr,
		"usage: %s [-v] [-c count] [-w wait] [-M core] [-N system ] [interface]\n",
		progname);
	exit(1);
}

/*
 * Called if an interval expires before intpr has completed a loop.
 * Sets a flag to not wait for the alarm.
 */
void
catchalarm()
{
	signalled = 1;
}

/*
 * Print a running summary of interface statistics.
 * Repeat display every interval seconds, showing statistics
 * collected over that interval.  Assumes that interval is non-zero.
 * First line printed at top of screen is always cumulative.
 */
void
intpr()
{
	STRUCT 	*sc, *osc;
	off_t 	addr;
	int 	oldmask, line = 0;

	addr = nl[N_SOFTC].n_value + unit * sizeof(STRUCT);
	sc = (STRUCT *) malloc(sizeof(STRUCT));
	osc = (STRUCT *) malloc(sizeof(STRUCT));
	memset(osc, 0, sizeof(STRUCT));

	while (1) {
		if (kvm_read(kd, addr, sc,
			     sizeof(STRUCT)) != sizeof(STRUCT))
			errx(1, "reading statistics: %s", kvm_geterr(kd));
		
		(void)signal(SIGALRM, catchalarm);
		signalled = 0;
		(void)alarm(interval);

		if ((line % 20) == 0) {
			printf("%8.8s %6.6s %6.6s %6.6s %6.6s",
				"IN", "PACK", "COMP", "UNCOMP", "ERR");
			if (vflag)
				printf(" %6.6s %6.6s", "TOSS", "IP");
			printf(" | %8.8s %6.6s %6.6s %6.6s %6.6s",
				"OUT", "PACK", "COMP", "UNCOMP", "IP");
			if (vflag)
				printf(" %6.6s %6.6s", "SEARCH", "MISS");
			putchar('\n');
		}
		printf("%8u %6d %6u %6u %6u",
		       V(sc_if.if_ibytes),
		       V(sc_if.if_ipackets),
		       V(sc_comp.sls_compressedin),
		       V(sc_comp.sls_uncompressedin),
		       V(sc_comp.sls_errorin));
		if (vflag)
			printf(" %6u %6u",
			       V(sc_comp.sls_tossed),
			       V(sc_if.if_ipackets) -
			       V(sc_comp.sls_compressedin) -
			       V(sc_comp.sls_uncompressedin) -
			       V(sc_comp.sls_errorin));
		printf(" | %8u %6d %6u %6u %6u",
			V(sc_if.if_obytes),
			V(sc_if.if_opackets),
			V(sc_comp.sls_compressed),
			V(sc_comp.sls_packets) - V(sc_comp.sls_compressed),
			V(sc_if.if_opackets) - V(sc_comp.sls_packets));
		if (vflag)
			printf(" %6u %6u",
				V(sc_comp.sls_searches),
				V(sc_comp.sls_misses));

		putchar('\n');
		fflush(stdout);
		line++;

		count--;
		if (!infinite && !count)
			break;
		oldmask = sigblock(sigmask(SIGALRM));
		if (signalled == 0)
			sigpause(0);
		sigsetmask(oldmask);
		signalled = 0;
		(void)alarm(interval);
		memcpy(osc, sc, sizeof(STRUCT));
	}
}

int
main(argc, argv)
	int argc;
	char *argv[];
{
	char 	errbuf[_POSIX2_LINE_MAX], tmp[IFNAMSIZ];
	struct 	ifreq ifr;
	char 	*kernel, *core;
	int 	c, s;

	kernel = core = NULL;
	strcpy(interface, INTERFACE_PREFIX);
	strcat(interface, "0");
	if ((progname = strrchr(argv[0], '/')) == NULL)
		progname = argv[0];
	else
		++progname;

	while ((c = getopt(argc, argv, ":c:vw:M:N:")) != -1) {
		switch (c) {
		case 'c':
			count = atoi(optarg);
			if (count <= 0)
				usage();
			break;
		case 'v':
			++vflag;
			break;
		case 'w':
			interval = atoi(optarg);
			if (interval <= 0)
				usage();
			break;
		case 'M':
			core = optarg;
			break;
		case 'N':
			kernel = optarg;
			break;
		case ':':
			fprintf(stderr,
				"option -%c requires an argument\n", optopt);
			usage();
			break;
		case '?':
			fprintf(stderr, "unrecognized option: -%c\n", optopt);
			usage();
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (!interval && count)
		interval = 5;
	if (interval && !count)
		infinite = 1;
	if (!interval && !count)
		count = 1;

	if (argc > 1)
		usage();
	if (argc > 0) {
		if (strlen(argv[0]) > sizeof(interface))
			errx(1, "invalid interface specified %s", interface);
		strcpy(interface, argv[0]);
	}

	strcpy(tmp, INTERFACE_PREFIX);
	strcat(tmp, "%d");
	if (sscanf(interface, tmp, &unit) != 1)
		errx(1, "invalid interface '%s' specified", interface);
	s = socket(AF_INET, SOCK_DGRAM, 0);
	if (s < 0)
		err(1, "creating socket");
	strcpy(ifr.ifr_name, interface);
	if (ioctl(s, SIOCGIFFLAGS, (caddr_t)&ifr) < 0)
		errx(1, "unable to confirm existence of interface '%s'",
		    interface);
	close(s);

	kd = kvm_openfiles(kernel, core, NULL, O_RDONLY, errbuf);
	if (kd == NULL)
		errx(1, "%s", errbuf);
	if (kvm_nlist(kd, nl) != 0)
		errx(1, "%s", kvm_geterr(kd));

	intpr();
	exit(0);
}
